plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+        -- Special case: if the new group and the accumulated term are both letnonrecs,
+        -- it merges them together into a single let-group (i.e. linear scoped).
+        genLetsFromScc :: Term tyname name uni a
+                       -> S.Set PLC.Unique
+                       -> State [S.Set PLC.Unique] (Term tyname name uni a)
+        genLetsFromScc accTerm scc = do
+              res <- forM (S.toList scc) (\ v ->
+                                             case M.lookup v rhsTable of
+                                               Just rhs@(_,_,_,oldDepth) ->
+                                                 -- lift each resulting binding in a list monoid to accumulate them
+                                                 -- visit the generated rhs-term as well for any potential floating
+                                                 (_3.bindingSubterms) (goTerm
+                                                                        -- inside the RHS we "pretend" that we are at the depth of the let in the original program,
+                                                                        -- since the depths of lets in FloatData correspond to the original depths.
+                                                                        oldDepth
+                                                                        -- for optimization, we pass only a part of the floatdata that are larger than this RHS orig. depth.
+                                                                        (snd $ IM.split oldDepth restDepthTable)
+                                                                      ) rhs
+                                               _ -> error "Something went wrong: no rhs was found for this let in the rhstable."
+                                                       )
+              let (mAnn, mRecurs, newBindings) = foldr1 (<>) $ fmap (\(oAnn,oRec,oBinding,_) -> (oAnn
+                                                                                     -- if the SCC is a single node then use its original 'Recursivity';
+                                                                                     -- otherwise, the SCC is  a group of nodes and we *have to* treat all of them in a 'letrec',
+                                                                                     -- since we don't have any information on how to linearize those.
+                                                                                   , if isSingleton scc then oRec else Rec
+                                                                                     -- lift the binding into a semigroup for accumulation
+                                                                                   , pure oBinding)
+                                                                    ) res
+              pure $ case accTerm of
```

Comment:
There's a `mkLet` function in `MkPir`. I propose we add this behaviour to it - it'll get a monoid constraint on `a`, but that's okay.

Reply:
@michaelpj I am not sure that this will improve the code, because I am already using NonEmpty  (lists) and NonEmptySets. It feels like defeating the whole purpose of using the "safer" nonempty and nonemptyset.

Reply:
I think this was in the context of us maybe ending up with normal lists somehow? I've forgotten now.